remote_attestation/rust/src/crypto.rs

```diff
@@ -55,17 +57,25 @@ impl aead::NonceSequence for OneNonceSequence {
 
 /// Implementation of Authenticated Encryption with Associated Data (AEAD).
 /// https://datatracker.ietf.org/doc/html/rfc5116
+///
+/// This implementation uses separate keys for encrypting data and decrypting peer encrypted data.
+/// Which means that this implementation uses the same key for encryption, which peer uses for
+/// decryption.
+/// It is necessary to prevent the Loopback Attack, where malicious network takes an outgoing packet
+/// and feeds it back as an incoming packet.
 pub struct AeadEncryptor {
-    /// Key used for encrypting and decrypting data.
-    ///
-    /// AES-GCM algorithm uses the same key for both encryption and decryption.
-    /// https://datatracker.ietf.org/doc/html/rfc5288
-    key: Vec<u8>,
+    /// Key used for encrypting data.
+    encryption_key: Vec<u8>,
```

Comment:
Is it possible to avoid storing these keys here, and just keep the sealing key? In fact, the current code looks like it is re-creating the sealing key every time encrypt is called, which does not seem ideal.

Reply:
First we create an `let unbound_sealing_key = aead::UnboundKey::new`.
But then this key is consumed by `let mut sealing_key = ring::aead::SealingKey::new(unbound_sealing_key,`

The problem is that the sealing key needs a Nonce (which is taken from the message every time, for decryption).
And for encryption, if we provide nonce to the key, I'm not sure we can extract each individual nonce after, in order to put it into Protobuf.

Reply:
Why not use a simple nonce sequence then, since that seems the expected underlying API? I thought a counter is sufficient. If not, then I guess we would need a nonce sequence to which we can feed the next nonce, extracted from the incoming message, just before calling encrypt or decrypt (though it may look a bit weird because of the extra synchronization). What does TLS use for nonce?

Reply:
Currently we are using random nonces, because this is done in Tink (with no way of changing it).

TLS (at least 1.3) is much more complicated in this sense:
```
It uses instead a combination of a fixed per-connection IV (derived at the same time as the key) and the sequence number, XORed - not concatenated.
```

Reply:
Note that the TLS approach is actually simpler to implement than the one you have here, since it can be encapsulated in a struct that implements the https://briansmith.org/rustdoc/ring/aead/trait.NonceSequence.html trait quite nicely (it just needs to have the IV, and then keep track of the counter internally). Anyways, let's leave that for later.